hilti/runtime/src/fiber.cc

```diff
@@ -54,66 +60,171 @@ void detail::_Trampoline(void* argsp) {
-        ::fiber_switch(fiber->_fiber.get(), fiber->_caller);
+// Captures arguments passed into stack switcher trampoline.
+struct SwitchArgs {
+    ::Fiber* switcher = nullptr;
+    detail::Fiber* from = nullptr;
+    detail::Fiber* to = nullptr;
+};
 
-        fiber->_finishSwitchFiber("trampoline-loop");
+// Entry point for stack switcher trampoline.
+void fiber_switch_trampoline(void* argsp) {
+    auto* args = reinterpret_cast<SwitchArgs*>(argsp);
+    auto from = args->from;
+    auto to = args->to;
+
+    auto from_tag = (from->isMain() ? "main" : "fiber");
+    auto to_tag = (to->isMain() ? "main" : "fiber");
+    HILTI_RT_DEBUG("fibers", fmt("[stack-switcher] switching from %s-%p to %s-%p", from_tag, from, to_tag, to));
+
+    auto* context = context::detail::get();
+    assert(context->fiber.current == nullptr);
+    context->fiber.current = to;
+
+    if ( ! from->isMain() ) {
+        // Copy old stack out.
+        from->saved_stack.region = detail::StackRegion(from->_fiber.get());
+        from->saved_stack.buffer = ::realloc(from->saved_stack.buffer, from->saved_stack.region.size());
+        if ( ! from->saved_stack.buffer )
+            throw RuntimeError("out of memory when saving fiber stack");
+
+        HILTI_RT_DEBUG("fibers", fmt("[stack-switcher] saving stack %s to %p", from->saved_stack.region,
+                                     from->saved_stack.buffer));
+        ::memcpy(from->saved_stack.buffer, from->saved_stack.region.lower, from->saved_stack.region.size());
     }
 
-    HILTI_RT_DEBUG("fibers", fmt("[%p] finished trampoline loop", fiber));
+    if ( ! to->isMain() && to->saved_stack.buffer ) {
+        // Copy new stack in.
+        HILTI_RT_DEBUG("fibers", fmt("[stack-switcher] restoring stack %s from %p", to->saved_stack.region,
+                                     to->saved_stack.buffer));
+        ::memcpy(to->saved_stack.region.lower, to->saved_stack.buffer, to->saved_stack.region.size());
+    }
+
+    ::fiber_switch(args->switcher, to->_fiber.get());
 }
+#endif
+}
+
+detail::FiberContext::FiberContext() {
+    main = std::make_unique<detail::Fiber>(true);
+    current = main.get();
+
+    switch_trampoline = std::make_unique<::Fiber>();
+    if ( ! ::fiber_alloc(switch_trampoline.get(), IndividualStackSize, fiber_bottom_abort, this,
+                         FIBER_FLAG_GUARD_LO | FIBER_FLAG_GUARD_HI) )
+        throw RuntimeError("could not allocate fiber switch trampoline");
+
+#ifdef USE_SHARED_STACK
+    // Instantiate an unused fiber just to create the shared stack.
+    shared_stack = std::make_unique<::Fiber>();
+    if ( ! ::fiber_alloc(shared_stack.get(), SharedStackSize, fiber_bottom_abort, this,
+                         FIBER_FLAG_GUARD_LO | FIBER_FLAG_GUARD_HI) )
+        throw RuntimeError("could not allocate shared stack");
+#endif
 }
 
-detail::Fiber::Fiber() : _fiber(std::make_unique<::Fiber>()), _caller(context::detail::get()->current_fiber) {
+detail::FiberContext::~FiberContext() {}
```

Comment:
Let's get rid of this if it isn't needed; else let's document what this works around.